File: Miscellaneous/Golang/101. Symmetric Tree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {

	root := &TreeNode{1, nil, nil}
	root.Left = &TreeNode{2, nil, nil}
	root.Left.Left = &TreeNode{3, nil, nil}
	root.Left.Right = &TreeNode{4, nil, nil}
	root.Right = &TreeNode{2, nil, nil}
	root.Right.Left = &TreeNode{4, nil, nil}
	root.Right.Right = &TreeNode{3, nil, nil}
	result := isSymmetric(root)
	fmt.Println(result)
}

func isSymmetric(root *TreeNode) bool {

	return isMirror(root, root)
}

func isMirror(root1, root2 *TreeNode) bool {
	if root1 == nil && root2 == nil {
		return true
	}

	if root1 != nil && root2 != nil {
		if root1.Val == root2.Val {
			return (isMirror(root1.Left, root2.Right)) && (isMirror(root1.Right, root2.Left))
		}
	}

	return false
}

File: Miscellaneous/Python/101. Symmetric Tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isMirror(root1, root2):
    if root1 is None and root2 is None:
        return True
    
    if root1 is not None and root2 is not None:
        if root1.val == root2.val:
            return isMirror(root1.left, root2.right) and isMirror(root1.right, root2.left)
    
    return False

class Solution:
    def isSymmetric(self, root):
        return isMirror(root, root)
        

obj = Solution()
root = TreeNode(1)
root.left = TreeNode(2)
root.left.left = TreeNode(3)
root.left.right = TreeNode(4)
root.right = TreeNode(2)
root.right.left = TreeNode(4)
root.right.right = TreeNode(3)
ans = obj.isSymmetric(root)
print(ans)